app/services/reply_scraper.rb

```diff
@@ -28,116 +28,126 @@ def initialize(reply, console: false)
+    created_at = comment.at_css('.datetime').text
+
     @reply.content = strip_content(content)
     @reply.editor_mode = 'html'
     @reply.created_at = @reply.updated_at = created_at
 
-    set_from_username(@reply, username)
-    set_from_icon(@reply, img_url, img_keyword)
-
-    if @reply.is_a?(Post)
-      @reply.last_user_id = @reply.user_id
-      @reply.edited_at = created_at
-    end
+    @reply.user = set_from_username(username)
+    @reply.icon = set_from_icon(img_url, img_keyword) if img_url
+    post_setup if @reply.is_a? Post
 
     Audited.audit_class.as_user(@reply.user) do
       @reply.save!
     end
   end
 
-  def set_from_username(tag, username)
-    if BASE_ACCOUNTS.key?(username)
-      tag.user = User.find_by(username: BASE_ACCOUNTS[username])
-      return
-    end
+  private
 
-    character = Character.find_by(screenname: username.tr("-", "_")) || Character.find_by(screenname: username.tr("_", "-"))
-    unless character
+  def post_setup
+    @reply.last_user_id = @reply.user_id
+    @reply.edited_at = @reply.created_at
+  end
+
+  def set_from_username(username)
+    return User.find_by(username: BASE_ACCOUNTS[username]) if BASE_ACCOUNTS.key?(username)
+
+    unless (character = Character.find_by(screenname: [username.tr("-", "_"), username.tr("_", "-")]))
       user = prompt_for_user(username)
       character = Character.create!(user: user, name: username, screenname: username)
       gallery = Gallery.create!(user: user, name: username)
       CharactersGallery.create!(character_id: character.id, gallery_id: gallery.id)
     end
 
-    tag.character = character
-    tag.user = character.user
+    @reply.character = character
+    character.user
   end
 
   def prompt_for_user(username)
     raise UnrecognizedUsernameError.new("Unrecognized username: #{username}") unless @console_import
-    print('User ID or username for ' + username + '? ')
+    print("User ID or username for #{username}? ")
     input = STDIN.gets.chomp
-    return User.find_by_id(input) if input.to_s == input.to_i.to_s
-    User.where('lower(username) = ?', input.downcase).first
+    return User.find_by(id: input) if input.to_s == input.to_i.to_s
+    User.find_by('lower(username) = ?', input.downcase)
   end
 
-  def set_from_icon(tag, url, keyword)
-    return unless url
+  def set_from_icon(url, keyword)
+    url = parse_url(url)
+    icon = Icon.find_by(url: url)
+    return icon if icon
+
+    keyword = parse_keyword(keyword)
+
+    if @reply.character
+      icon = @reply.character.icons.find_by(keyword: keyword)
+      icon ||= clean_keyword(keyword)
+      return icon if icon
+    end
 
-    url = 'https://v.dreamwidth.org' + url if url[0] == '/'
-    host_url = url.gsub(/https?:\/\//, "")
-    https_url = 'https://' + host_url
-    icon = Icon.find_by(url: https_url)
-    tag.icon = icon and return if icon
+    create_icon(url, keyword)
+  end
 
+  def parse_url(url)
+    uri = URI(url)
+    uri.host ||= 'v.dreamwidth.org'
```

Comment:
Looks like we also want to set the scheme to `https` (rather than http / blank), not sure we're doing that anymore